Sources/AsyncHTTPClient/FileDownloadDelegate.swift

```diff
@@ -128,4 +177,14 @@ public final class FileDownloadDelegate: HTTPClientResponseDelegate {
         self.finalize()
         return self.progress
     }
+
+    deinit {
+        switch threadPool {
+        case .unowned:
+            break
+        case .owned(let pool):
+            // if the delegate is unused we still need to shutdown the thread pool
+            try! pool.syncShutdownGracefully()
```

Comment:
we shouldn't destroy this in `deinit`, you don't even have a real guarantee that `deinit` isn't running in a thread owned by the pool (which would make it deadlock).

I'd suggest to add a lifecycle hook (shutdown or so) to the delegate and have AHC invoke that. That new shutdown method can have an empty default implementation to not break the API.

Separately, I think AHC should _provide_ the thread pool or else we spin up/down a thread per download and that kinda makes it pointless to do asynchronous I/O if we need a new thread each time.

Reply:
actually you could add two delegate methods (with default empty impls):

- `provideThreadPool(ioPool)` where AHC provides a running thread pool to the delegate
- `shutdown()` where AHC tells the delegate that it its job is done

After that, we can change the default here and have the delegate not even have the option to accept an io pool from the user.

Reply:
If AHC provides the default shared thread pool we don't need the `shutdown()` lifecycle hook because the delegate doesn't need to shutdown the thread pool anymore but the AHC client.

Reply:
that's true

Reply:
I have added a shared file IO thread pool to the `HTTPClient` which is passed to the delegate through a call to `provideSharedThreadPool(fileIOPool:)`
https://github.com/swift-server/async-http-client/pull/614/commits/1afc151d644c489a233916d26b99b966d6154dee